src/cross_toml.rs

```diff
@@ -67,6 +90,60 @@ impl CrossToml {
         Ok((cfg, unused))
     }
 
+    /// Merges another [`CrossToml`] into `self` and returns a new merged one
+    ///
+    /// # Merging of `targets`
+    /// The `targets` entries are merged based on the [`Target`] keys.
+    /// If a [`Target`] key is present in both configs, the [`CrossTargetConfig`]
+    /// in `other` overwrites the one in `self`.
+    ///
+    /// # Merging of `build`
+    /// The `build` fields ([`CrossBuildConfig`]) are merged based on their sub-fields.
+    /// A field in the [`CrossBuildConfig`] will only overwrite another if it contains
+    /// a value, i.e. it is not `None`.
+    pub fn merge(self, other: CrossToml) -> Result<CrossToml> {
+        type ValueMap = serde_json::Map<String, serde_json::Value>;
+
+        fn to_map<S: Serialize>(s: S) -> Result<ValueMap> {
+            if let Some(obj) = serde_json::to_value(s)?.as_object() {
+                Ok(obj.to_owned())
+            } else {
+                panic!("Failed to serialize CrossToml as object");
+            }
+        }
+
+        fn from_map<D: DeserializeOwned>(map: ValueMap) -> Result<D> {
+            let value = serde_json::to_value(map)?;
+            Ok(serde_json::from_value(value)?)
+        }
+
+        // Builds maps of objects
+        let mut self_map = to_map(&self)?;
+        let other_map = to_map(other)?;
+        let cfg_fields: Vec<String> = self_map.keys().cloned().collect();
+
+        // Iterates over and merges config fields
+        for field in cfg_fields.iter() {
+            let self_sub_map = self_map.get_mut(field).unwrap().as_object_mut().unwrap();
+            let mut other_sub_map = other_map
+                .get(field)
+                .cloned()
+                .unwrap()
+                .as_object_mut()
+                .unwrap()
+                .to_owned();
+
+            // Only retain fields in the build config that have a value
+            if field == "build" {
```

Comment:
I think we can generalize this to also include `env`, and also handle arrays, since we might want an array value in `build` in the future or an optional value in `env` later on. That is, this should be:

```rust
if field == "build" || field == "env" {
    other_sub_map.retain(|_, v| !v.is_null());
}
```

Reply:
I've made the merging logic more generalizable [here](https://github.com/Alexhuszagh/cross/blob/cargo_toml_overwrite_arrays/src/cross_toml.rs#L120-L154), so it doesn't assume the exact structure of our config file which will make this code easier to maintain.

```rust
        // merge 2 objects. y has precedence over x.
        fn merge_objects(x: &mut ValueMap, y: &ValueMap) -> Option<()> {
            // we need to iterate over both keys, so we need a full deduplication
            let keys: BTreeSet<String> = x.keys().chain(y.keys()).cloned().collect();
            for key in keys {
                let in_x = x.contains_key(&key);
                let in_y = y.contains_key(&key);
                if !in_x && in_y {
                    let yk = y[&key].clone();
                    x.insert(key, yk);
                    continue;
                } else if !in_y {
                    continue;
                }


                let xk = x.get_mut(&key)?;
                let yk = y.get(&key)?;
                if xk.is_null() && !yk.is_null() {
                    *xk = yk.clone();
                    continue;
                } else if yk.is_null() {
                    continue;
                }


                // now we've filtered out missing keys and optional values
                // all key/value pairs should be same type.
                if xk.is_object() {
                    merge_objects(xk.as_object_mut()?, yk.as_object()?)?;
                } else {
                    *xk = yk.clone();
                }
            }


            Some(())
        }
```

I think this is an improvement since it will work no matter how our config file is structured, no matter how many tables or fields we add.